test/e2e/network/service.go

```diff
@@ -2310,6 +2311,115 @@ var _ = common.SIGDescribe("Services", func() {
+		framework.ExpectNoError(err)
+		nodeCounts := len(nodes.Items)
+		if nodeCounts < 2 {
+			e2eskipper.Skipf("The test requires at least two ready nodes on %s, but found %v", framework.TestContext.Provider, nodeCounts)
+		}
+		node0 := nodes.Items[0]
+
+		serviceName := "svc-proxy-terminating"
+		ns := f.Namespace.Name
+		servicePort := 80
+
+		ginkgo.By("creating a TCP service " + serviceName + " where all pods are terminating" + ns)
+		jig := e2eservice.NewTestJig(cs, ns, serviceName)
+		svc, err := jig.CreateTCPService(func(svc *v1.Service) {
+			svc.Spec.Ports = []v1.ServicePort{
+				{Port: 80, Name: "http", Protocol: v1.ProtocolTCP, TargetPort: intstr.FromInt(80)},
+			}
+			svc.Spec.Type = v1.ServiceTypeLoadBalancer
+			svc.Spec.ExternalTrafficPolicy = v1.ServiceExternalTrafficPolicyTypeLocal
+		})
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Creating 1 webserver pod to be part of the TCP service")
+		webserverPod0 := e2epod.NewAgnhostPod(ns, "echo-hostname-0", nil, nil, nil, "netexec", "--http-port", strconv.Itoa(servicePort), "--delay-shutdown", "600")
+		webserverPod0.Labels = jig.Labels
+		webserverPod0.Spec.TerminationGracePeriodSeconds = utilpointer.Int64(600)
+		e2epod.SetNodeSelection(&webserverPod0.Spec, e2epod.NodeSelection{Name: node0.Name})
+
+		_, err = cs.CoreV1().Pods(ns).Create(context.TODO(), webserverPod0, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+		framework.ExpectNoError(e2epod.WaitTimeoutForPodReadyInNamespace(f.ClientSet, webserverPod0.Name, f.Namespace.Name, framework.PodStartTimeout))
+		validateEndpointsPortsOrFail(cs, ns, serviceName, portsByPodName{webserverPod0.Name: {servicePort}})
+
+		pausePod0 := e2epod.NewAgnhostPod(ns, "pause-pod-0", nil, nil, nil)
+		e2epod.SetNodeSelection(&pausePod0.Spec, e2epod.NodeSelection{Name: node0.Name})
+
+		pausePod0, err = cs.CoreV1().Pods(ns).Create(context.TODO(), pausePod0, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+		framework.ExpectNoError(e2epod.WaitTimeoutForPodReadyInNamespace(f.ClientSet, pausePod0.Name, f.Namespace.Name, framework.PodStartTimeout))
+
+		// validate that the health check node port from kube-proxy returns 200 when there are ready endpoints
+		nodeIPs := e2enode.GetAddresses(&node0, v1.NodeInternalIP)
+		healthCheckNodePortAddr := net.JoinHostPort(nodeIPs[0], strconv.Itoa(int(svc.Spec.HealthCheckNodePort)))
+		err = wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
+			cmd := fmt.Sprintf(`curl -q -s --connect-timeout 5 http://%s/healthz`, healthCheckNodePortAddr)
+			out, err := framework.RunHostCmd(pausePod0.Namespace, pausePod0.Name, cmd)
+			if err != nil {
+				return false, err
+			}
+
+			sanitizedJson := &bytes.Buffer{}
+			if err := json.Compact(sanitizedJson, []byte(out)); err != nil {
+				return false, err
+			}
+
+			expectedOut := fmt.Sprintf(`{"service":{"namespace":%q,"name":%q},"localEndpoints":%d}`, svc.Namespace, svc.Name, 1)
+			if sanitizedJson.String() != expectedOut {
```

Comment:
So while HealthCheckNodePort success/failure is a required detail of any service proxy. I'm not sure that we require all service proxies to output exactly the same json as kube-proxy does. I'd make the test only check whether the result is 200 or 503.

Reply:
I'm not sure we can also rely on all service proxies to reliably use the 503 status code, but agreed it's better than checking the actual response body, will update the test accordingly.